### PR TITLE
Pipeline support

### PR DESCRIPTION
**Pull request by ArieShout**
Add Jenkins Pipeline support for the service-fabric plugin. The `Action`'s were removed, as they didn't provide much useful function as I can see. 

* Deploy with Azure resource configuration:

   ```groovy
   azureServiceFabricPublish azureCredentialsId: 'Jenkins-credentials-ID-for-Azure-service-principal',
                             resourceGroup: 'resource-group-containing-the-service-fabric-cluster',
                             serviceFabric: 'service-fabric-name',
                             clientCert: 'Certificate-File-Path',
                             clientKey: 'Key-File-Path',
                             applicationName: 'fabric:/Your-Application-Name',
                             applicationType: 'Your-Application-Type',
                             manifestPath: 'Relative/path/to/ApplicationManifest.xml'
   ```

* Deploy with direct Service Fabric endpoint:

   ```groovy
   azureServiceFabricPublish clusterHost: 'Host.name.of.your.cluster',
                             clientCert: 'Certificate-File-Path',
                             clientKey: 'Key-File-Path',
                             applicationName: 'fabric:/Your-Application-Name',
                             applicationType: 'Your-Application-Type',
                             manifestPath: 'Relative/path/to/ApplicationManifest.xml'
   ```

The core change was refactoring the build logic into `ServiceFabricPublishStep` which provides the functionality for the Pipeline execution, and reference it from `ServiceFabricPublisher` which provides the functionality for the normal build configured in the web GUI. Some background about the refactoring:

We need to run the Shell script to manipulate the Service Fabric deployment with `sfctl`. There are two implementation of the Shell script build for Jenkins:

* `hudson.tasks.Shell` used in the original implementation, which provides the **Execute shell** build in the web GUI
* `org.jenkinsci.plugins.workflow.steps.durable_task.ShellStep` which provides the `sh` function in the Pipeline

However, the two implementation are not interchangeable. `Shell` requires an `AbstractBuild` which is provided in the job configured with the web GUI, but it is not available in Pipeline execution. `ShellStep` requires a `StepContext` which is provided in Pipeline execution, but it's not available in normal build.

In order to cope with this, I extracted the core logic to `ServiceFabricPublishStep` which will serve the Pipeline function `azureServiceFabricPublish` with `ServiceFabricPublishStep#start`. The original `ServiceFabricPublisher` just reference the step implementation, and provide support for the normal build configured from the web GUI.

### CODE REVIEW

**Comment by suhuruli** [comment]
@rapatchi  
@raunakpandya

File: src/main/java/com/microsoft/jenkins/servicefabric/ServiceFabricPublishStep.java [hunk 1]
```diff
@@ -0,0 +1,418 @@
...
+                    Constants.AI_RUN, buildId,
+                    "Subscription", AppInsightsUtils.hash(azure.subscriptionId()),
+                    "ResourceGroup", AppInsightsUtils.hash(resourceGroup),
+                    "Cluster", AppInsightsUtils.hash(serviceFabric));
+
+            ServiceFabricCluster cluster = new ServiceFabricCluster(azure, resourceGroup, serviceFabric);
+            String managementEndpoint = cluster.getManagementEndpoint();
+            try {
+                URL url = new URL(managementEndpoint);
+                if ("https".equalsIgnoreCase(url.getProtocol())) {
```
Comment (rapatchi) [inline src/main/java/com/microsoft/jenkins/servicefabric/ServiceFabricPublishStep.java]: Is it possible to check this before we run this pipeline step ?. Like doing a validation as done by "doFillServiceFabricItems" or "doCheckApplicationName".

Comment (ArieShout) [reply src/main/java/com/microsoft/jenkins/servicefabric/ServiceFabricPublishStep.java]: The `doCheckXXX` method only receives the value being checked (in this case, the Service Fabric resource name), which is not enough for us to retrieve the Service Fabric details.

We can add a "Verify" button and implement customized verification logic which pulls all the required configuration fields, and and do deep verification if user clicks the button. We may check the following items in this case:

* whether the service fabric is secured, and if yes, do we have certificate & key configured
* can we connect to the service fabric cluster

This is not a trivial change, so I suggest we do that in a separate PR.

Comment (rapatchi) [reply src/main/java/com/microsoft/jenkins/servicefabric/ServiceFabricPublishStep.java]: Ok Sure.

File: README.md [hunk 2]
```diff
@@ -16,8 +16,39 @@ To manually install the plugin:
...
+## Pipeline
+
+You can call the function `azureServiceFabricPublish` from your Jenkins Pipeline to deploy your application to
+Service Fabric. You can also use the the Jenkins ***Pipeline Syntax*** to generate the Pipeline snippet from
+the form UI.
+
+* Deploy with Azure resource configuration:
+
+   ```groovy
+   azureServiceFabricPublish azureCredentialsId: 'Jenkins-credentials-ID-for-Azure-service-principal',
```
Comment (suhuruli) [inline README.md]: @rapatchi @raunakpandya @ArieShout 

How are we able to connect to a cluster using Service Principles? Are you parsing the endpoint from the return somewhere?

Comment (ArieShout) [reply README.md]: The logic used to connect to service fabric cluster was unchanged. I implemented the logic to fetch the service fabric details from Azure, using the service principal, resource group and cluster name provided.

**Pull request merged**